### ISSUE TITLE
The`#[serde(flatten)]` syntax is not supported with the feature `arbitrary precision`

### ISSUE THREAD

**Issue by koushiro** [serde-rs/json#721]
example:

```rust
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct Inner {
    inner: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Outer {
    out: u32,
    #[serde(flatten)]
    inner: Inner,
}

fn main() {
    let out = Outer {
        out: 1,
        inner: Inner {
            inner: 2,
        }
    };
    let ser = serde_json::to_string(&out).unwrap();
    println!("Out: {}", ser);
    let de: Outer = serde_json::from_str(&ser).unwrap();
    println!("Out: {:?}", de);
}
```

with the feature `arbitrary_precision`:
```toml
[dependencies]
serde = { version = "1.0", features = ["derive"] }
serde_json = { version = "1.0", features = ["arbitrary_precision"] }
```

result:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/json-demo`
Out: {"out":1,"inner":2}
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: map, expected u32", line: 1, column: 19)', src/main.rs:24:48
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

without the feature `arbitrary_precision`:
```toml
[dependencies]
serde = { version = "1.0", features = ["derive"] }
serde_json = { version = "1.0" }
```

result:

```
    Finished dev [unoptimized + debuginfo] target(s) in 2.39s
     Running `target/debug/json-demo`
Out: {"out":1,"inner":2}
Out: Outer { out: 1, inner: Inner { inner: 2 } }
```

**Comment by koushiro** [comment]
@dtolnay Can you give me some tips on how to fix this issue?

**Comment by jerry73204** [comment]
It does not work with `f64` either. It fails even without `arbitrary_precision`. Suppose it is a bug of `serde(flatten)`?

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Inner {
        pub rotation: [f64; 3],
        pub translation: [f64; 3],
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    struct Outer {
        #[serde(flatten)]
        inner: Inner,
    }

    #[test]
    fn wtf() -> Result<()> {
        let text = r#"{
  "rotation": [
    0.7547604535512583,
    -0.12463161307955047,
    0.10464618142566347,
    0.6354941348976522
  ],
  "translation": [
    -0.07889085071772557,
    -0.05290491930288589,
    -0.04170572461992302
  ]
}
"#;
        let _: Outer = serde_json::from_str(&text)?;

        Ok(())
    }
}
```

**Comment by tylerhawkes** [comment]
I just ran into this. I have two `#[serde(flatten)]` fields that were working and it took me a couple hours to figure out that it was the addition of the arbitrary precision feature by another dependency that was causing my tests to fail.

**Comment by ewoolsey** [comment]
> I just ran into this. I have two #[serde(flatten)] fields that were working and it took me a couple hours to figure out that it was the addition of the arbitrary precision feature by another dependency that was causing my tests to fail.

Yeah also just ran into this in my project. It can be a pretty sinister bug to track down. I'm honestly surprised this issue is still open considering how code breaking it can be. Took me and a coworker like 5 hours combined to track it down.

**Comment by xbtmatt** [comment]
Just ran into this bug, note that it is fairly difficult to track down because the bug only exists if you try to deserialize directly with `serde_json::from_str` to your final type.

If you do an intermediate deserialization to a `serde_json::Value` it works just fine, even with the offending feature flag + `#[serde(flatten)]` field combo.

Works regardless of feature/flattened fields:

```rust
let value: serde_json::Value =
    serde_json::from_str(&json).unwrap();
let my_type: YourType =
    serde_json::from_value(value).unwrap();
```

Fails with `invalid type: map, expected f64` (or whatever type you're using) if `arbitrary_precision` is enabled:

```rust
let my_type: YourType  =
    serde_json::from_str(&json).unwrap();
```

So either go deserialize into `serde_json::Value` then to your type, or make sure transitive deps aren't pulling in the arbitrary precision feature.

**Comment by puellavulnerata** [comment]
The original example in this issue **passes now**, which makes the issue look stale — it isn't. The integer case was fixed, the general case wasn't, and the discriminator turns out to be the JSON *token* rather than the Rust type.

`{"out":1,"inner":2}` started working in **1.0.75**, per that release's note:

> Fix deserialization of small integers in arbitrary_precision mode (#845)

Bisecting the issue's own example against real releases: it fails on ≤ 1.0.74 with exactly the reported `invalid type: map, expected u32`, and passes from 1.0.75 onward.

## Still broken on 1.0.151

```rust
use serde::Deserialize;

#[derive(Deserialize)]
struct Inner {
    n: f64,
}

#[derive(Deserialize)]
struct Outer {
    #[serde(flatten)]
    inner: Inner,
}

fn main() {
    // Integer token: fine since 1.0.75.
    println!("{:?}", serde_json::from_str::<Outer>(r#"{"n":2}"#).map(|o| o.inner.n));

    // Any non-integer token: still fails.
    println!("{:?}", serde_json::from_str::<Outer>(r#"{"n":2.0}"#).map(|o| o.inner.n));
    println!("{:?}", serde_json::from_str::<Outer>(r#"{"n":1.5}"#).map(|o| o.inner.n));
    println!("{:?}", serde_json::from_str::<Outer>(r#"{"n":1e2}"#).map(|o| o.inner.n));
}
```

```toml
[dependencies]
serde = { version = "1.0.229", features = ["derive"] }
serde_json = { version = "1.0.151", features = ["arbitrary_precision"] }
```

Output:

```
Ok(2.0)
Err(Error("invalid type: map, expected f64", line: 1, column: 9))
Err(Error("invalid type: map, expected f64", line: 1, column: 9))
Err(Error("invalid type: map, expected f64", line: 1, column: 9))
```

## It's the token, not the type

The first and second lines above deserialize into the **same** `f64` field and differ only in how the number is written. Holding the target at `f64` and varying only the literal:

| JSON token | result |
|---|---|
| `2` | ok |
| `-2` | ok |
| `9007199254740993` | ok |
| `2.0` | **fail** |
| `0.0` | **fail** |
| `1.5` | **fail** |
| `1e2` | **fail** |
| `2E+3` | **fail** |
| `123456789012345678901234567890` | **fail** |

So the 1.0.75 fast path covers number tokens that fit an `i64`/`u64`; anything with a `.`, an exponent, or too many digits for a `u64` still reaches the visitor as the internal map and fails against any typed field.

That also explains why it reads as type-dependent in the reports above — an integer-typed field can only ever receive an integer token, so integer fields look fixed while float fields look broken. A float field receiving `2` works fine, and an integer-shaped-but-huge token fails.

Same picture for `f32`, `Option<f64>`, `Vec<f64>` and `[f64; 3]`: all fail on a non-integer token, all fine on an integer one.

## Unchanged from the earlier comments

`from_value` still works in every case above, so the two-step `from_str::<Value>` → `from_value::<T>` workaround holds. And the failure is deserialization-only — serialization is fine throughout.

Verified on rustc 1.98.0, serde 1.0.229, serde_json 1.0.151.
